aggregator/election_mgr.go

```diff
@@ -0,0 +1,501 @@
+	electionKey      string
+	electionWg       sync.WaitGroup
+	electionState    ElectionState
+	changeInProgress bool
+	changeCancelFn   context.CancelFunc
+	changeWg         sync.WaitGroup
+	sleepFn          sleepFn
+	metrics          electionManagerMetrics
+}
+
+// NewElectionManager creates a new election manager.
+func NewElectionManager(opts ElectionManagerOptions) ElectionManager {
+	instrumentOpts := opts.InstrumentOptions()
+	campaignOpts := opts.CampaignOptions()
+	campaignRetrier := xretry.NewRetrier(opts.CampaignRetryOptions().SetForever(true))
+	changeRetrier := xretry.NewRetrier(opts.ChangeRetryOptions().SetForever(true))
+	mgr := &electionManager{
+		nowFn:           opts.ClockOptions().NowFn(),
+		logger:          instrumentOpts.Logger(),
+		reportInterval:  instrumentOpts.ReportInterval(),
+		campaignOpts:    campaignOpts,
+		electionOpts:    opts.ElectionOptions(),
+		campaignRetrier: campaignRetrier,
+		changeRetrier:   changeRetrier,
+		electionKeyFmt:  opts.ElectionKeyFmt(),
+		leaderService:   opts.LeaderService(),
+		leaderValue:     campaignOpts.LeaderValue(),
+		state:           electionManagerNotOpen,
+		resignCh:        make(chan struct{}),
+		doneCh:          make(chan struct{}),
+		electionState:   FollowerState,
+		sleepFn:         time.Sleep,
+		metrics:         newElectionManagerMetrics(instrumentOpts.MetricsScope()),
+	}
+	go mgr.reportMetrics()
+	return mgr
+}
+
+func (mgr *electionManager) Open(shardSetID string) error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerNotOpen {
+		return errElectionManagerExpectNotOpen
+	}
+	mgr.state = electionManagerOpen
+	mgr.electionKey = fmt.Sprintf(mgr.electionKeyFmt, shardSetID)
+	mgr.electionWg.Add(1)
+	go mgr.startElectionLoop()
+	return nil
+}
+
+func (mgr *electionManager) ElectionState() ElectionState {
+	mgr.RLock()
+	electionState := mgr.electionState
+	mgr.RUnlock()
+	return electionState
+}
+
+func (mgr *electionManager) Resign(ctx context.Context) error {
+	mgr.Lock()
+	if mgr.state != electionManagerOpen {
+		mgr.Unlock()
+		return errElectionManagerExpectOpen
+	}
+	mgr.state = electionManagerResigning
+	mgr.resignWg = &sync.WaitGroup{}
+	resignWg := mgr.resignWg
+	resignWg.Add(1)
+	mgr.Unlock()
+
+	err := mgr.leaderService.Resign(mgr.electionKey)
+	if err != nil {
+		resignWg.Done()
+		mgr.Lock()
+		mgr.state = electionManagerOpen
+		mgr.resignWg = nil
+		mgr.Unlock()
+		err := fmt.Errorf("leader service resign failed: %v", err)
+		mgr.metrics.resignErrors.Inc(1)
+		mgr.logError("resign failed", err)
+		return err
+	}
+
+	// Wait for the election loop to exit after resigning.
+	close(mgr.resignCh)
+	resignWg.Done()
+	mgr.electionWg.Wait()
+
+	// Now wait for the current instance to become a follower.
+	var (
+		ctxDone          = false
+		isFollower       = false
+		stateCheckPeriod = time.Second
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			ctxDone = true
+		default:
+			isFollower = mgr.ElectionState() == FollowerState
```

Comment:
Might it be better to have some signal for follower state instead of a sleep and check here?